### PR TITLE
GH-121439: Allow PyTupleObjects with an ob_size of 20 in the free_list to be reused

### PR DESCRIPTION
**Pull request by satori1995**
In the `maybe_freelist_push` function, PyTupleObjects with an ob_size of 20 are being stored in the free_list for potential reuse. However, in the `maybe_freelist_pop` function, these objects aren't actually being reused.

* Issue: gh-121439

### CODE REVIEW

**Comment by eendebakpt** [comment]
Nice catch! Can you create an issue for this change? See https://devguide.python.org/getting-started/pull-request-lifecycle/#quick-guide

**Comment by satori1995** [comment]
> Nice catch! Can you create an issue for this change? See https://devguide.python.org/getting-started/pull-request-lifecycle/#quick-guide

Sure, I'll create an issue for this change.

**Comment by eendebakpt** [comment]
@satori1995 Thanks! Could you also write a news item using the blurb tool?

**Comment by satori1995** [comment]
@eendebakpt All done

File: Misc/NEWS.d/next/Core and Builtins/2024-07-08-02-24-55.gh-issue-121439.jDHod3.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Allow PyTupleObjects with an ob_size of 20 in the free_list to be reused
```
Comment (eendebakpt) [inline Misc/NEWS.d/next/Core and Builtins/2024-07-08-02-24-55.gh-issue-121439.jDHod3.rst]:
```suggestion
Allow `PyTupleObjects` with an `ob_size` of 20 in the `free_list` to be reused
```

**Review by eendebakpt — APPROVED** [review]
Looks good! Tuples with size 20 are probably not very common, but since the freelist is allocated for them let's use it when possible.

File: Misc/NEWS.d/next/Core and Builtins/2024-07-08-02-24-55.gh-issue-121439.jDHod3.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Allow `PyTupleObjects` with an `ob_size` of 20 in the `free_list` to be reused
```
Comment (eendebakpt) [inline Misc/NEWS.d/next/Core and Builtins/2024-07-08-02-24-55.gh-issue-121439.jDHod3.rst]:
```suggestion
Allow ``PyTupleObjects`` with an ``ob_size`` of 20 in the ``free_list`` to be reused
```

**Comment by eendebakpt** [comment]
> @eendebakpt Hi, When will this submission be merged? Is there anything else I need to do?

One of the CI tests is still failing (formatting of the news entry). If that is fixed all CI checks should be green.

The PR needs to be reviewed and approved by a core developer in order to be merged. It can take some time before this happens though, as there are many issues and PRs requiring attention.

**Review by corona10 — CHANGES_REQUESTED** [review]
Overall changes looks good to me but please fix the linter issue.
https://github.com/python/cpython/actions/runs/9836190774/job/27151370914?pr=121428

**Comment by satori1995** [comment]
@corona10 Sorry，I have made the requested changes，but  I don't understand why there are still three checks failing. Just to mention, this is my first time contributing code to CPython, so I'm not very clear on some of the details.

**Review by satori1995 — COMMENTED** [review]
I have made the requested changes

**Comment by eendebakpt** [comment]
@satori1995 Failing docs test could be due do the formatting in the news entry, although I am not sure as the error message does not show any line our text of the news entry in this PR. Could you rebase your branch onto latest dev? (in github there is a button "Update branch" for this). Maybe that resolves the failing test.

File: Misc/NEWS.d/next/Core and Builtins/2024-07-08-02-24-55.gh-issue-121439.jDHod3.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Allow ``PyTupleObjects`` with an ``ob_size`` of 20 in the ``free_list`` to be reused
```
Comment (eendebakpt) [inline Misc/NEWS.d/next/Core and Builtins/2024-07-08-02-24-55.gh-issue-121439.jDHod3.rst]:
```suggestion
Allow :c:type:`PyTupleObject` with an :c:member:`~PyTypeObject.ob_size` of 20 in the freelist of tuples to be reused.
```
@satori1995 I am no expert in formatting of the news entries, so this is a bit of trial and error and looking at other news entries

**Comment by satori1995** [comment]
I don’t know how to make all the checks pass, so I give up.😂

**Comment by eendebakpt** [comment]
@satori1995 The failing checks we can resolve, your part of the PR looks good. Could you sync your PR the the main branch? (the Update branch button)

**Review by corona10 — APPROVED** [review]
lgtm

**Comment by satori1995** [comment]
Thank you for your help, @corona10 has already merged the changes into the main branch. 

Here, I'd like to share how I discovered this issue. A few years ago, I read the CPython 3.8 source code, but recently, I have had more free time, so I started reading the CPython 3.12 source code. When I looked at the parts related to tuple, I found that tuples of length 20 were being cached but not reused.

To verify my conclusion, I tried modifying CPython by adding a classmethod to <class 'tuple'>.
`static PyObject *get_free_list_count(PyObject *cls, Py_ssize_t len)`

When the tuple length is between 1 and 19, deleting a tuple increases the count returned by `get_free_list_count` by 1, and creating a tuple decreases the count by 1. But when the tuple length is 20, deleting a tuple increases the count by 1, but creating a tuple does not change the count.

Therefore, I concluded that this is a bug, and this bug has not been fixed in version 3.13 either. So I attempted to submit my changes. However, since this was my first time submitting code to CPython and I was not familiar with the related procedures, the whole process was a bit difficult.

Here, I sincerely thank @eendebakpt and @corona10 for their patience, help, and guidance. May your lives always be filled with happiness.

**Comment by corona10** [comment]
@satori1995 Backport patches are automatically handled.
- https://github.com/python/cpython/pull/121565
- https://github.com/python/cpython/pull/121566

**Comment by satori1995** [comment]
> @satori1995 Backport patches are automatically handled.
> 
> * [[3.13] GH-121439: Allow PyTupleObjects with an ob_size of 20 in the free_list to be reused (gh-121428) #121565](https://github.com/python/cpython/pull/121565)
> * [[3.12] GH-121439: Allow PyTupleObjects with an ob_size of 20 in the free_list to be reused (gh-121428) #121566](https://github.com/python/cpython/pull/121566)

👌🏻ok

**Pull request merged**